README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Cross-Entropy Optimisation of Noisy Functions

## Presentation

The package `noisyCE2` implements the cross-entropy algorithm
(Rubinstein and Kroese, 2004) for the optimisation of unconstrained
deterministic and noisy functions through a highly flexible and
customisable function which allows user to define custom variable
domains, sampling distributions, updating and smoothing rules, and
stopping criteria. Several built-in methods and settings make the
package very easy-to-use under standard optimisation problems.

## Three examples

### Negative paraboloid

The negative 4-dimensional paraboloid can be maximised as follows:

``` r
negparaboloid <- function(x) { -sum((x - (1:4))^2) }

sol <- noisyCE2(negparaboloid, domain = rep('real', 4))
```

### Rosenbrock’s function

The 10-dimensional Rosenbrock’s function can be minimised as follows:

``` r
rosenbrock <- function(x) {
  sum(100 * (tail(x, -1) - head(x, -1)^2)^2 + (head(x, -1) - 1)^2)
}

newvar <- type_real(
  init = c(0, 2),
  smooth = list(
    quote(smooth_lin(x, xt, 1)),
    quote(smooth_dec(x, xt, 0.7, 5))
  )
)

sol <- noisyCE2(
  rosenbrock, domain = rep(list(newvar), 10),
  maximise = FALSE, N = 2000, maxiter = 10000
)
```

### Noisy negative paraboloid

The negative 4-dimensional paraboloid with additive Gaussian noise can
be maximised as follows:

``` r
noisyparaboloid <- function(x) { -sum((x - (1:4))^2) + rnorm(1) }

sol <- noisyCE2(noisyparaboloid, domain = rep('real', 4), stoprule = geweke(x))
```

where the stopping criterion based on the Geweke’s test has been adopted
according to Bee et al. (2017).

## References

Bee M., G. Espa, D. Giuliani, F. Santi (2017) “A cross-entropy approach
to the estimation of generalised linear multilevel models”, *Journal of
Computational and Graphical Statistics*, **26** (3), pp. 695-708.
<https://doi.org/10.1080/10618600.2016.1278003>

Rubinstein, R. Y., and Kroese, D. P. (2004), *The Cross-Entropy Method*,
Springer, New York. ISBN: 978-1-4419-1940-3


NEWS.md


# noisyCE2 1.1.0


## Major changes

* Redefined the default value of argument `stopwindow` of function `noisyCE`
  according to Bee et al. (2017).


## Minor changes

* Added methods `summary` and `coef` for `S3` class `noisyCE2`.
* Added recordings of computation times for `noisyCE2`.
* Included `testthat` among suggested packages.


## Typos and other marginal changes

* Completed references information.
* Corrected syntax on `return`.





# noisyCE2 1.0.0

First release of the package `noisyCE2` on CRAN!




tests/testthat.R
library(testthat)
library(noisyCE2)

test_check("noisyCE2")


tests/testthat/test_stoprules.R
library(testthat)
library(noisyCE2)



test_that('ts_change works for tol argument', {
  expect_identical(as.numeric(ts_change(rnorm(1))), 1)
  expect_identical(as.numeric(ts_change(c( 100,  104), tol = 3)), 1)
  expect_identical(as.numeric(ts_change(c(-100, -104), tol = 3)), 1)
  expect_identical(as.numeric(ts_change(c( 100,  103), tol = 3)), 1)
  expect_identical(as.numeric(ts_change(c( 100, -103), tol = 3)), 1)
  expect_identical(as.numeric(ts_change(c( 100,  102), tol = 3)), 0)
  expect_identical(as.numeric(ts_change(c( 100, -102), tol = 3)), 1)
})


test_that('ts_change works for reltol argument', {
  expect_identical(as.numeric(ts_change(rnorm(1))), 1)
  expect_identical(as.numeric(ts_change(c( 100,  104), reltol = 0.03)), 1)
  expect_identical(as.numeric(ts_change(c(-100, -104), reltol = 0.03)), 1)
  expect_identical(as.numeric(ts_change(c( 100,  103), reltol = 0.03)), 1)
  expect_identical(as.numeric(ts_change(c( 100, -103), reltol = 0.03)), 1)
  expect_identical(as.numeric(ts_change(c( 100,  102), reltol = 0.03)), 0)
  expect_identical(as.numeric(ts_change(c( 100, -102), reltol = 0.03)), 1)
})


test_that('ts_change works for arguments tol and reltol', {
  expect_identical(as.numeric(ts_change(c(   0,    2), tol = 3)), 0)
  expect_identical(as.numeric(ts_change(c(   0,   -3), tol = 3)), 1)
  expect_identical(as.numeric(ts_change(c( 100,  101), tol = 2, reltol = 0.03)), 0)
  expect_identical(as.numeric(ts_change(c( 100,  102), tol = 2, reltol = 0.03)), 0)
  expect_identical(as.numeric(ts_change(c( 100,  102), tol = 4, reltol = 0.01)), 0)
  expect_identical(as.numeric(ts_change(c( 100,  104), tol = 2, reltol = 0.03)), 1)
})




R/typevar.R

#' Functions for defining the types of variables
#' 
#' All functions permit fully-customised types of variable to be defined.
#' Functions other than `type_custom` already include standard default values
#' which make the definition of standard variable types easier and quicker.
#'
#' @param type label for identifying the type of variable. The name is not
#'   used internally in any case.
#' @param init `numeric` vector of starting values of parameters of the
#'   sampling distribution.
#' @param randomXj function for randomly generating variable values according
#'   to the sampling distribution. The function should take the number of
#'   observations to be generated as a first argument, and the vector of
#'   parameters as a second argument; a vector of random values should be
#'   returned.
#' @param x2v function for updating the parameters of the sampling distribution.
#'   *No smoothing is needed.* The function should take a single argument to be
#'   used for updating the parameters.
#' @param v2x function for obtaining point values of variable from the
#'   parameters of the sampling distribution.
#' @param smooth list of unevaluated expressions of smoothing functions for each
#'   parameter of the sampling distribution.
#' @param ... further arguments to be included into the `typevar` object. In
#'   case of function for predefined types, it is possible to use ellipsis for
#'   overwriting default values (see § Examples).
#'
#' @return
#' An object of class `type` and `typevar`, where `type` is the value of the
#' argument `type` passed to `type_custom`, or predefined lables (if not
#' overwritten) in case of other functions.
#'
#' @examples
#' # Define a new type of real variable where the first parameter of the
#' # sampling distribution is updated through the median (instead of the
#' # mean):
#' type_real(
#'   type = 'real2', 
#'   x2v = function(x) { c(median(x), sd(x)) }
#' )
#'
#' # Define a new type of real variable whith different smoothing
#' # parameters:
#' type_real(
#'   type = 'real3', 
#'   smooth = list(
#'     quote(smooth_lin(x, xt, 0.8)),
#'     quote(smooth_dec(x, xt, 0.99, 15))
#'   )
#' )
#'
#' @name type_variable
#'
#' @export
type_custom <- function(type = 'custom', init = c(0, 10),
  randomXj = function(n, v) { rnorm(n, v[1], v[2]) },
  x2v = function(x) { c(mean(x), sd(x)) }, v2x = function(v) { v[1] },
  smooth = list(quote(smooth_lin(x, xt, 1)), quote(smooth_dec(x, xt, 0.9, 10))), ...) {
  	
  # Set explicit arguments
  list(
    type = type,
    init = init,
    randomXj = randomXj,
    x2v = x2v,
    v2x = v2x
  ) %>%
    # Include user's changes
    modifyList(list(...)) -> out
     
  if (is.null(out[['smooth']])) { out$smooth <- smooth }
  
  # Set classes
  structure(out, class = c(out$type, 'typevar'))
}



#' @rdname type_variable
#' @export
type_real <- function(...) {
  # Set default values
  list(
    type = 'real',
    init = c(0, 10),
    randomXj = function(n, v) { rnorm(n, v[1], v[2]) },
    x2v = function(x) { c(mean(x), sd(x)) },
    v2x = function(v) { v[1] }
  ) %>%
    # Include user's changes
    modifyList(list(...)) -> out
    
  if (is.null(out[['smooth']])) {
  	out$smooth <- list(
      quote(smooth_lin(x, xt, 1)),
      quote(smooth_dec(x, xt, 0.9, 10))
    )
  }
  
  # Set classes
  structure(out, class = c(out$type, 'typevar'))
}



#' @rdname type_variable
#' @export
type_positive <- function(...) {
  # Set default values
  list(
    type = 'positive',
    init = c(0, 10),
    randomXj = function(n, v) { rlnorm(n, v[1], v[2]) },
    x2v = function(x) { log(x) %>% { c(mean(.), sd(.)) } %>% return() },
    v2x = function(v) { exp(v[1] + 0.5 * v[2]^2) }
  ) %>%
    # Include user's changes
    modifyList(list(...)) -> out
    
  if (is.null(out[['smooth']])) {
  	out$smooth <- list(
      quote(smooth_lin(x, xt, 1)),
      quote(smooth_dec(x, xt, 0.9, 10))
    )
  }
  
  # Set classes
  structure(out, class = c(out$type, 'typevar'))}



#' @rdname type_variable
#' @export
type_negative <- function(...) {
  # Set default values
  list(
    type = 'negative',
    init = c(0, 10),
    randomXj = function(n, v) { -rlnorm(n, v[1], v[2]) },
    x2v = function(x) { log(-x) %>% { c(mean(.), sd(.)) } %>% return() },
    v2x = function(v) { -exp(v[1] + 0.5 * v[2]^2) }
  ) %>%
    # Include user's changes
    modifyList(list(...)) -> out
    
  if (is.null(out[['smooth']])) {
  	out$smooth <- list(
      quote(smooth_lin(x, xt, 1)),
      quote(smooth_dec(x, xt, 0.9, 10))
    )
  }
  
  # Set classes
  structure(out, class = c(out$type, 'typevar'))
}




R/noisyCE2.R

#' Cross-Entropy Optimisation of Noisy Functions
#' 
#' Unconstraint optimisation of noisy functions through the cross-entropy
#' algorithm.
#'
#' @param f objective function which takes the vector of optimisation variables
#'   as first argument.
#' @param domain a `list` (or other coercible objects) where each component
#'   specifies the domain of each variable of the objective function `f`.
#'   The components of the list may be either objects of `typevar` class (see
#'   [type_variable]) or strings identifying one of [type_variable] functions
#'   (for example `"real"` for function [type_real()]). See § Examples.
#' @param ... other arguments to be passed to `f` or to other methods (for
#'   `print` and `plot`).
#' @param rho parameter \eqn{\rho} of the Cross-Entropy algorithm. This argument
#'   may be passed either as a numeric value in \eqn{(0,1)} or as an unevaluated
#'   expression which may include the number of current iteration `n`, or the
#'   argument `N`.
#' @param N parameter \eqn{N} of the Cross-Entropy algorithm. This argument
#'   may be passed either as a positive integer or as an unevaluated expression
#'   which may include the number of current iteration `n`.
#' @param smooth list of unevaluated expressions to be used as smoothing rules
#'   for the parameters of the sampling probability distributions of **all
#'   variables**. If not `NULL`, all default or set smoothing rules of all
#'   variables will be overwritten. See [type_variable] for details and examples.
#' @param stopwindow unevaluated expression returning the object to be passed to
#'   the stopping rule. Symbol `gam` permits the time series \eqn{\gamma_t} to
#'   be used (as a `numeric` vector).
#' @param stoprule stopping rule passed as an unevaluated expression including
#'   `x` as the object returned by evaluation of argument `stopwindow`. The
#'   algorithm is stopped when zero is returned by the evaluation of `stoprule`.
#'   If returned object has attribute `mess`, this is used as a message.
#'   Currently, built-in stopping rules are [ts_change()] and [geweke()], others
#'   may be defined by user.
#' @param maxiter maximum number of iteration. When it is reached, algorithm is
#'   stopped whether or not the stopping criterion is satisfied. If the maximum
#'   number of iteration is reached, the `code` and the `message` components of
#'   `noisyCE` object are overwritten.
#' @param maximise if `TRUE` (default) `f` is maximised, otherwise a
#'   minimisation of `f` is performed.
#' @param x,object object of class `noisyCE2`, as returned by `noisyCE2`.
#' @param verbose algorithm verbosity (values `v`, `vv` and `vvv` are admitted).
#' @param what type of plot should be drawn. If `what = "x"` (default), values
#'   of the variables are plotted as time series; if `what = "gam"`, time series
#'   of statistics \eqn{\gamma} is plotted; if `what = "param"`, time series of
#'   parameters of the sampling distributions are plotted.
#' @param start,end first and last value to be plotted. If `NULL`, all values
#'   are plotted.
#'
#' @return
#' An object of class `noisyCE2` structured as a list with the following
#' components:
#' \item{f}{argument `f`.}
#' \item{fobj}{objective function `f` where possible arguments passed through
#'   argument `...` have been substituted. Thus, the value of the objective
#'   function maximised by `noisyCE` in `x0` can be computed as `fobj(x0)`. If
#'   a minimisation has been performed, `fobj` returns `f` with sign inverted.}
#' \item{xopt}{`numeric` vector with solution.}
#' \item{hxopt}{matrix of `niter` rows and `length(xopt)` columns with values of
#'   variables generated by the optimisation algorithm.}
#' \item{param}{`list` of `length(xopt)` components where time series of
#'   parameters (vectors \eqn{v_t}) are stored for each variable as `data.frame`
#'   objects with `niter+1` rows (the first rows are the starting values set
#'   through function `noisyCEcontrol`).}
#' \item{gam}{vector of values \eqn{\gamma_t}.}
#' \item{niter}{number of iterations.}
#' \item{code}{convergence code of the algorithm. Value `0` means that algorithm
#'   has converged; other values are defined according to the stopping rule.}
#' \item{convMess}{textual message associated to the convergence code (if any).}
#' \item{compTimes}{named vector computation times of each phase.}
#'
#' @examples
#' library(magrittr)
#' # Optimisation of the 4-dimensional function:
#' # f(x1,x2,x3,x4)=-(x1-1)^2-(x2-2)^2-(x3-3)^2-(x4-4)^2
#' sol <- noisyCE2(function(x) -sum((x - (1:4))^2), domain = rep('real', 4))
#' # Representation of the convergence process:
#' plot(sol, what = 'x')
#' plot(sol, what = 'gam')
#'
#' @export
noisyCE2 <- function(f, domain, ..., rho = 0.05, N = 1000,
  smooth = NULL,
  stopwindow = tail(gam, (n > 20) * n / 2), stoprule = ts_change(x),
  maxiter = 1000, maximise = TRUE, verbose = 'v') {
  
  # Read date and time
  compTimes <- Sys.time()
  
  # Redefine the objective function including the other parameters
  fobj <- function(x) { (2 * maximise - 1) * f(x, ...) }
  
  # Save variable names
  namesX <- names(domain)
  if (is.null(namesX)) {
  	namesX <- paste0('X', seq_along(domain))
  }
  
  # Define the block of variables
  if (is.null(smooth)) {
  	block <- generate_block(domain)
  } else {
  	block <- generate_block(domain, smooth = smooth)
  }
  
  # Initialisations
  n <- 0
  vt <- lapply(block, function(x) { x$init })
  vhist <- lapply(vt, function(x) as.data.frame(matrix(x, 1)))
  gam <- NULL
  convCode <- -2
  
  # Cross-entropy algorithm
  compTimes %<>% c(Sys.time())
  
  while((convCode != 0) & (n < maxiter)) {
  	
    # Update the counter
    n %<>% add(1)

    # Simulate the N values of the independent variables
    X <- mapply(
      FUN = function(varj, theta) { varj$randomXj(eval(N), theta) },
      block, vt, SIMPLIFY = FALSE
    )

    # Simulate the N values of the objective function
    X %<>% Reduce(cbind, ., NULL)
    fx <- apply(X, 1, fobj)
    
    # Compute and store the new gamma
    fx %>%
      quantile(1 - eval(rho), type = 1, na.rm = TRUE) %>%
      c(gam, .) -> gam
    
    # Update the parameters
    X[fx >= gam[n], ] %>%
      data.frame %>%
      mapply(
        FUN = function(varj, xj) { varj$x2v(xj) },
        block, ., SIMPLIFY = FALSE
      ) -> vt

    # Smooth the parameters
    xt <- NULL # Only set for avoiding NOTE in package check
    vt %<>% smooth_block(block, ., vhist)
    vhist %<>% concat_block_vect(vt)
    
    # Progress message
    vmessage(verbose, 1,
      '\r     * iter. ',
      format(n, digits = 4, nsmall = 0, width = 4),
      ' - gamma: ',
      format(gam[n], digits = 4, nsmall = 4),
      '       ',
      sep = '', appendLF = FALSE
    )
    
    # Check the stopping criterion
    x <- eval(substitute(stopwindow))
    if (length(x) > 0) {
      convCode <- unname(eval(substitute(stoprule)))
    }
  }
  
  compTimes %<>% c(Sys.time())
  
  if((convCode != 0) & (n >= maxiter)) {
  	convCode <- -1
  	attr(convCode, 'convMess') <- 'the maximum number of iterations has been reached'
  }
  
  # Output
  mess <- attr(convCode, 'convMess')
  vmessage(verbose, 1, '\n     * The algorithm has ', mess)

  #  
  mapply(
    FUN = function(varj, vj) { unlist(varj$v2x(vj)) },
    block, vhist, SIMPLIFY = FALSE
  ) %>%
    Reduce(cbind, .) %>%
    data.frame %>%
    set_rownames(NULL) -> hxopt

  if(any(is.null(namesX))) {
    colnames(hxopt) <- paste0('X', 1:ncol(hxopt))	
  } else {
  	colnames(hxopt) <- namesX
  }
  
  tail(hxopt, 1) %>%
    as.numeric %>%
    set_names(colnames(hxopt)) -> xopt
  
  vhist %<>% set_names(namesX)
  
  #
  compTimes %<>%
    c(Sys.time()) %>%
    diff %>%
    set_names(c('Initialisation', 'Optimisation', 'Preparing output'))
  
  # Output
  list(
    f = f, fobj = fobj, xopt = xopt, hxopt = hxopt, param = vhist,
    gam = unname(gam), niter = n, code = convCode, convMess = mess,
    compTimes = compTimes
  ) %>%
    structure(class = 'noisyCE2') %>%
    return()
}



#' @describeIn noisyCE2 display synthetic information about a `noisyCE2` object
#' @export
print.noisyCE2 <- function(x, ...) {
  cat('Object of class "noisyCE2"\n')
  cat('--------------------------------------\n')
  cat('Status               :', x$convMess, '\n')
  cat('Computation time     :', 
    format(unclass(sum(x$compTime)), digits = 4),
    attr(x$compTime, 'units'), '\n'
  )
  cat('Number of iterations :', x$niter, '\n')
  cat('Time per iteration   :',
    format(unclass(sum(x$compTime) / x$niter), digits = 4),
    attr(x$compTime / x$niter, 'units'), '\n'
  )
  cat('Last gamma value     :', tail(x$gam, 1), '\n')
  cat('Solution x*          :', x$xopt, '\n')
  #cat('Optimal value f(x*)  :', x$fobj(x$xopt), '\n')
  invisible(x)
}



#' @describeIn noisyCE2 display summary information about a `noisyCE2` object
#' @export
summary.noisyCE2 <- function(object, ...) {
  print(object, ...)
}



#' @describeIn noisyCE2 plot various components of a `noisyCE2` object
#' @export
plot.noisyCE2 <- function(x, what = c('x', 'gam', 'param'),
  start = NULL, end = NULL, ...) {
  	
  what %<>% match.arg(c('x', 'gam', 'param'))
  
  if (what == 'x') {
    stats::ts(x$hxopt, 1) %>%
      stats::window(start, end) %>%
      stats::plot.ts(...)
  }
  if (what == 'gam') {
    stats::ts(x$gam, 1) %>%
      stats::window(start, end) %>%
      stats::plot.ts(...)
  }
  if (what == 'param') {
    for(j in seq_along(x$param)) {
      colnames(x$param[[j]]) %<>% paste0(names(x$param)[j], '_', .)
    }
    x$param %>%
      Reduce(cbind, .) %>%
      stats::ts(0) %>%
      stats::window(start, end) %>%
      stats::plot.ts(...)
  }
  
  invisible(NULL)
}



#' @describeIn noisyCE2 get the solution of the optimisation
#' @export
coef.noisyCE2 <- function(object, ...) {
  object$xopt
}




R/stopping-rules.R

#' Geweke's test stopping rule
#' 
#' `geweke` tests the convergence of `x` through the Geweke's test.
#'
#' @param x `numeric` vector of last \eqn{\gamma_n} values, as selected by the
#'   function passed to [noisyCE2()] through the argument `stopwindow`.
#' @param frac1,frac2 fraction arguments of the Geweke's test according to
#'   [coda::geweke.diag()].
#' @param pvalue threshold of the \eqn{p}-value which triggers the stop of the
#'   algorithm.
#'
#' @return
#' A `numeric` indicating whether the algorithm has converged:
#' \item{0}{the algorithm has converged.}
#' \item{1}{the algorithm has not converged.}
#'
#' @family stopping rules
#'
#' @export
geweke <- function(x, frac1 = 0.3, frac2 = 0.4, pvalue = 0.05) {
  x %>%
    coda::as.mcmc() %>%
    coda::geweke.diag(frac1 = frac1, frac2 = frac2) %>%
    use_series('z') %>%
    abs %>%
    pnorm(lower.tail = FALSE) %>%
    is_greater_than(pvalue) %>%
    not %>%
    as.numeric -> out
    
  attr(out, 'convMess') <- 'converged'
  if(out != 0) {
  	attr(out, 'convMess') <- 'not converged'
  }
  
  return(out)
}



#' Time series change stopping rule
#' 
#' Deterministic stopping rule based on the last change in the value of
#' \eqn{\gamma_n}. Changes smaller than `tol`, or relative changes
#' smaller than `reltol` stop the algorithm. This criterion is suitable
#' only in case of deterministic objective functions.
#'
#' @inheritParams geweke
#' @param reltol relative changes smaller than `tol` stop the algorithm.
#' @param tol changes smaller than `tol` stop the algorithm.
#'
#' @inherit geweke return
#'
#' @family stopping rules
#'
#' @export
ts_change <- function(x, reltol = 1e-4, tol = 1e-12) {
  out <- 1
  
  if (length(x) > 1) {
    x %<>% tail(n = 2)
    dx <- abs(diff(x))
    rdx <- min(dx / abs(x[1]), reltol + 1, na.rm = TRUE)
    if ((dx < tol) | (rdx < reltol)) { out <- 0 }
  }
    
  attr(out, 'convMess') <- 'converged'
  if(out != 0) {
  	attr(out, 'convMess') %<>% paste('not', .)
  }
  
  return(out)
}




R/block-functions.R

# domain
generate_block <- function(domain, ...) {
  domain %>%
    as.list %>%
    lapply(function(x) {
      if (inherits(x, 'typevar')) {
      	depo <- list(...)
      	if (!is.null(depo[['smooth']])) { x[['smooth']] <- NULL }
      	x %<>%
      	  modifyList(depo) %>%
      	  do.call('type_custom', args = .)
      } else {
      	if (is.character(x)) {
      	  x <- do.call(paste0('type_', x), args = list(...))
      	} else {
      	  stop('domain not defined')
      	}
      }
      
      return(x)
    }) %>%
    structure(class = 'blockvar') %>%
    return()
}


# smooth
smooth_block <- function(block, vt, vhist) {
  mapply(
    FUN = function(varj, vtj, vhistj) {
      mapply(
        function(fsm, x, xt) { eval(fsm, list(x = x, xt = xt)) },
      	varj$smooth, as.list(vtj), vhistj
      )
    },
    block, vt, vhist, SIMPLIFY = FALSE
  )
}



# concatenate
concat_block_vect <- function(xhist, xnew) {
  mapply(
    FUN = function(x, y) { as.data.frame(rbind(x, y)) },
    xhist, xnew, SIMPLIFY = FALSE)
}






R/smoothing-rules.R


#' Linear first-order smoothing rule
#' 
#' Linear smoothing rule
#' \deqn{x_{t+1}:=a\,x_t + (1-a)\,x_{t-1}}
#' for some \eqn{a\in[0,1]}.
#'
#' @param x `numeric` value of the last value of the parameter.
#' @param xt `numeric` vector of past values of the parameter (time series).
#' @param a smoothing parameter \eqn{a}.
#'
#' @return
#' A `numeric` vector of updated parameters.
#'
#' @family smoothing rules
#'
#' @export
smooth_lin <- function(x, xt, a) {
  a * x + (1 - a) * tail(xt, n = 1)
}



#' Decreasing first-order smoothing rule
#' 
#' Decreasing smoothing rule
#' \deqn{x_{t+1}:=a_t\,x_t + (1-a_t)\,x_{t-1}}
#' where
#' \deqn{a_t:= b\,\left(1-\left(1-\frac{1}{t}\right)^q\right)}
#' for some \eqn{0.7\leq b\leq1} and some \eqn{5\leq q\leq10}.
#'
#' @inheritParams smooth_lin
#' @param b smoothing parameter \eqn{b}.
#' @param qu smoothing parameter \eqn{q}.
#'
#' @inherit smooth_lin return
#'
#' @family smoothing rules
#'
#' @export
smooth_dec <- function(x, xt, b, qu) {
  depo <- b * (1 - (1 - 1 / length(xt))^qu)
  smooth_lin(x, xt, depo)
}


  








R/noisyCE2-package.R
#' Cross-Entropy Optimisation of Noisy Functions
#'
#' The package `noisyCE2` implements the cross-entropy algorithm (Rubinstein and
#' Kroese, 2004) for the optimisation of unconstrained deterministic and noisy
#' functions through a highly flexible and customisable function which allows
#' user to define custom variable domains, sampling distributions, updating and
#' smoothing rules, and stopping criteria. Several built-in methods and settings
#' make the package very easy-to-use under standard optimisation problems.
#'
#'
#' The package permits a noisy function to be maximised by means of the
#' cross-entropy algorithm. Formally, problems in the form
#' \deqn{\max_{x\in\Theta}\textbf{E}(f(x))}
#' are tackled for a noisy function
#' \eqn{f\colon\Theta\subseteq\textbf{R}^m\to\textbf{R}}.
#'
#' @references
#' Bee M., G. Espa, D. Giuliani, F. Santi (2017) "A cross-entropy approach to
#' the estimation of generalised linear multilevel models", *Journal of
#' Computational and Graphical Statistics*, **26** (3), pp. 695-708.
#' <https://doi.org/10.1080/10618600.2016.1278003>
#'
#' Rubinstein, R. Y., and Kroese, D. P. (2004), *The Cross-Entropy Method*,
#' Springer, New York. ISBN: 978-1-4419-1940-3
#'
#' @examples
#' \donttest{
#' # EXAMPLE 1
#' # The negative 4-dimensional paraboloid can be maximised as follows:
#' negparaboloid <- function(x) { -sum((x - (1:4))^2) }
#' sol <- noisyCE2(negparaboloid, domain = rep('real', 4))
#'
#' # EXAMPLE 2
#' # The 10-dimensional Rosenbrock's function can be minimised as follows:
#' rosenbrock <- function(x) {
#'   sum(100 * (tail(x, -1) - head(x, -1)^2)^2 + (head(x, -1) - 1)^2)
#' }
#'
#' newvar <- type_real(
#'   init = c(0, 2),
#'   smooth = list(
#'     quote(smooth_lin(x, xt, 1)),
#'     quote(smooth_dec(x, xt, 0.7, 5))
#'   )
#' )
#' 
#' sol <- noisyCE2(
#'   rosenbrock, domain = rep(list(newvar), 10),
#'   maximise = FALSE, N = 2000, maxiter = 10000
#' )
#'
#' # EXAMPLE 3
#' # The negative 4-dimensional paraboloid with additive Gaussian noise can be
#' # maximised as follows:
#' noisyparaboloid <- function(x) { -sum((x - (1:4))^2) + rnorm(1) }
#' sol <- noisyCE2(noisyparaboloid, domain = rep('real', 4), stoprule = geweke(x))
#' # where the stopping criterion based on the Geweke's test has been adopted
#' # according to Bee et al. (2017).
#' }
#'
#' @encoding UTF-8
#' @aliases noisyCE2-package
#'
#' @import magrittr
#'
#' @importFrom stats pnorm dnorm quantile sd integrate rnorm rlnorm
#' @importFrom utils flush.console tail modifyList
"_PACKAGE"

utils::globalVariables('.')
